File: src/main/java/com/intetics/lukyanenko/dao/jdbc/JDBCDAOFactory.java
package com.intetics.lukyanenko.dao.jdbc;

import com.intetics.lukyanenko.dao.*;
import com.intetics.lukyanenko.models.*;

public class JDBCDAOFactory implements DAOFactory
{
  private final AppUserDAO appUserDAO;
  private final OrderDAO orderDAO;
  private final OrderDetailDAO orderDetailDAO;
  private final GoodsItemDAO goodsItemDAO;
  private final GoodsCategoryDAO goodsCategoryDAO;
  private final CustomerDAO customerDAO;
  
  public JDBCDAOFactory(
                         AppUserDAO appUserDAO,
                         OrderDAO orderDAO,
                         OrderDetailDAO orderDetailDAO,
                         GoodsItemDAO goodsItemDAO,
                         GoodsCategoryDAO goodsCategoryDAO,
                         CustomerDAO customerDAO
                       )
  {
    this.appUserDAO = appUserDAO;
    this.orderDAO = orderDAO;
    this.orderDetailDAO = orderDetailDAO;
    this.goodsItemDAO = goodsItemDAO;
    this.goodsCategoryDAO = goodsCategoryDAO;
    this.customerDAO = customerDAO;
  }
  
  public <T extends Common> CommonDAO<T> getDAO(Class<T> modelClass)
  {
    if (modelClass == Order.class)
      return (CommonDAO<T>)orderDAO;
    else if (modelClass == OrderDetail.class)
      return (CommonDAO<T>)orderDetailDAO;
    else if (modelClass == AppUser.class)
      return (CommonDAO<T>)appUserDAO;
    else if (modelClass == GoodsCategory.class)
      return (CommonDAO<T>)goodsCategoryDAO;
    else if (modelClass == GoodsItem.class)
      return (CommonDAO<T>)goodsItemDAO;
    else if (modelClass == Customer.class)
      return (CommonDAO<T>)customerDAO;
    else
      throw new IllegalArgumentException(String.format("Model class %s is not supported", modelClass.getSimpleName()));
  }
}

File: src/main/java/com/intetics/lukyanenko/dao/jdbc/JDBCDAOFactory.kt
package com.intetics.lukyanenko.dao.jdbc

import com.intetics.lukyanenko.dao.*
import com.intetics.lukyanenko.models.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import kotlin.reflect.KClass

@Service
class JDBCDAOFactory @Autowired
constructor(
        private val appUserDAO: AppUserDAO,
        private val orderDAO: OrderDAO,
        private val orderDetailDAO: OrderDetailDAO,
        private val goodsItemDAO: GoodsItemDAO,
        private val goodsCategoryDAO: GoodsCategoryDAO,
        private val customerDAO: CustomerDAO
) : DAOFactory {

    override fun <T : Common> getDAO(modelClass: KClass<T>): CommonDAO<T> {
        return when (modelClass) {
            Order::class -> orderDAO
            OrderDetail::class -> orderDetailDAO
            AppUser::class -> appUserDAO
            GoodsCategory::class -> goodsCategoryDAO
            GoodsItem::class -> goodsItemDAO
            Customer::class -> customerDAO
            else -> throw IllegalArgumentException(String.format("Model class %s is not supported", modelClass.simpleName))
        } as CommonDAO<T>
    }
}